### DISCUSSION TITLE
Behavior inconsistency between Android and iOS when a client loses connection

### DISCUSSION THREAD

**Discussion by vitorbaraujo** [libuv/libuv#3685] [Q&A]
I have a TCP server running on macOS and two mobile clients, one in iOS and another with Android. When I turn off wifi on the Android device, I get an instant response in the `uv_read_start` read callback with `nread` set to 53 (software caused connection abort). However, when doing the same on the iOS device, I don't get the instant response and need to wait for the client to time out so I can prompt the reconnection.

As a separate test, I've implemented a simple TCP client using raw sockets on iOS and when I turn off wifi I receive an instant error from the delegate.

These tests were done with libuv versions 1.23.0, 1.23.1 and 1.44.1.

Do you know if this inconsistency is expected or it's some kind of misconfiguration on any of the clients?

**Comment by bnoordhuis** [comment]
Can you elaborate on what you mean by "raw sockets"? Raw has a specific meaning in the context of sockets (forging IP headers, think `ping`) but I infer that's not what you mean.

**Reply by vitorbaraujo** [reply]
What I meant by that is that I implemented a TCP client using the `CoreFoundation.CFStream` API using Swift. The main idea is to create a read/write stream using [CFStreamCreatePairWithSocketToHost](https://developer.apple.com/documentation/corefoundation/1539739-cfstreamcreatepairwithsockettoho?language=objc) and then listen to events using  the [StreamDelegate](https://developer.apple.com/documentation/foundation/streamdelegate) delegate. When turning off the device wifi using this method I receive a [Stream.Event.errorOccurred](https://developer.apple.com/documentation/foundation/stream/event/1414980-erroroccurred) event on the delegate.

**Reply by bnoordhuis** [reply]
I speculate (can't check because closed source) CoreFoundation listens for network device events and reports "device down" as errors on affected sockets.
